dev/statement_serializer.h

```diff
@@ -45,81 +44,222 @@ namespace sqlite_orm {
-                return move(a) + sep[a.empty()] + v;
+                ss << sep[std::exchange(first, false)] << serialize_order_by(arg, context);
             });
+            return ss;
+        }
+
+        // serialize and stream a vector of expressions;
+        // comma-separated
+        template<class C, class Ctx>
+        std::ostream& operator<<(std::ostream& ss, std::tuple<decltype((streaming_dynamic_expressions)), C, Ctx> tpl) {
+            const auto& args = std::get<1>(tpl);
+            auto& context = std::get<2>(tpl);
+
+            constexpr const char* sep[] = {", ", ""};
+            for(size_t i = 0, first = true; i < args.size(); ++i) {
+                ss << sep[std::exchange(first, false)] << serialize(args[i], context);
+            }
+            return ss;
         }
 
+        // stream a vector of already serialized strings;
+        // comma-separated
         template<class C>
-        std::string join_serialized_strings(const C& c) {
-            return join_serialized_strings(c.cbegin(), c.cend());
+        std::ostream& operator<<(std::ostream& ss, std::tuple<decltype((streaming_serialized)), C> tpl) {
+            const auto& strings = std::get<1>(tpl);
+
+            constexpr const char* sep[] = {", ", ""};
+            for(size_t i = 0, first = true; i < strings.size(); ++i) {
+                ss << sep[std::exchange(first, false)] << strings[i];
+            }
+            return ss;
         }
 
-        template<class T, class SFINAE = void>
-        struct statement_serializer;
+        // stream an identifier described by a variadic string pack, which is one of:
+        // 1. identifier
+        // 2. identifier, alias
+        // 3. qualifier, identifier, alias
+        template<class... Strings>
+        std::ostream& operator<<(std::ostream& ss, std::tuple<decltype((streaming_identifier)), Strings...> tpl) {
+            stream_identifier(ss, tpl, streaming_identifier.offset_index(std::index_sequence_for<Strings...>{}));
+            return ss;
+        }
 
-        template<class T, class I>
-        std::string serialize(const T& t, const serializer_context<I>& context) {
-            statement_serializer<T> serializer;
-            return serializer(t, context);
+        // stream a container of identifiers described by a string or a tuple, which is one of:
+        // 1. identifier
+        // 1. tuple(identifier)
+        // 2. tuple(identifier, alias), pair(identifier, alias)
+        // 3. tuple(qualifier, identifier, alias)
+        //
+        // comma-separated
+        template<class C>
+        std::ostream& operator<<(std::ostream& ss, std::tuple<decltype((streaming_identifiers)), C> tpl) {
+            const auto& identifiers = std::get<1>(tpl);
+
+            constexpr const char* sep[] = {", ", ""};
+            bool first = true;
+            for(auto& identifier: identifiers) {
+                ss << sep[std::exchange(first, false)];
```

Comment:
this is a very fucking elegant code

Reply:
Thx! It's tight. The only little issue one might argue that the exchange operation involves a temporary variable.